File: salesforce/lld/connection-pool/python/pool/connection_pool.py
"""Thread-safe bounded connection pool."""

from __future__ import annotations

import threading
import time
from collections import deque
from collections.abc import Callable

from .connection import Connection


class PoolTimeoutError(RuntimeError):
    """Raised when borrow waits for the timeout but no connection is released."""


class InvalidResourceError(RuntimeError):
    """Raised for foreign resources and double releases."""


ConnectionFactory = Callable[[str], Connection]
ValidationStrategy = Callable[[Connection], bool]


def default_factory(connection_id: str) -> Connection:
    """Factory abstraction: callers can swap creation without changing the pool."""

    return Connection(connection_id)


def default_validation(connection: Connection) -> bool:
    """Validation-on-borrow strategy; fake connections are valid while open."""

    return connection.is_valid()


class ConnectionPool:
    """Bounded, thread-safe generic resource pool.

    The MVP eagerly creates ``max_size`` connections, so allocation is capped by construction. A
    ``threading.Condition`` owns both the free deque and the borrowed set: ``borrow`` waits while the
    deque is empty, and ``release`` appends then notifies exactly one waiter.
    """

    def __init__(
        self,
        max_size: int,
        factory: ConnectionFactory = default_factory,
        validation_strategy: ValidationStrategy = default_validation,
    ) -> None:
        if max_size <= 0:
            raise ValueError("max_size must be positive")
        self._max_size = max_size
        self._validation = validation_strategy
        self._available: deque[Connection] = deque()
        self._all_ids: set[int] = set()
        self._borrowed_ids: set[int] = set()
        self._lock = threading.Lock()
        self._not_empty = threading.Condition(self._lock)

        for i in range(1, max_size + 1):
            connection = factory(f"conn-{i}")
            self._available.append(connection)
            self._all_ids.add(id(connection))

    def borrow(self, timeout: float) -> Connection:
        """Borrow a connection, blocking up to ``timeout`` seconds.

        The condition wait handles back-pressure: when all resources are checked out, callers sleep
        instead of spinning. Popping from ``_available`` and marking ``_borrowed_ids`` happen under
        the same lock, so a connection cannot be handed to two callers at once.
        """

        deadline = time.monotonic() + timeout
        with self._not_empty:
            while not self._available:
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise PoolTimeoutError("Timed out waiting for a connection")
                self._not_empty.wait(remaining)
            connection = self._available.popleft()
            self._borrowed_ids.add(id(connection))

        if not self._validation(connection):
            self.release(connection)
            raise InvalidResourceError(f"Borrowed connection is not valid: {connection.id}")
        return connection

    def release(self, connection: Connection) -> None:
        """Return a borrowed connection to the pool.

        Identity checks reject a connection created outside this pool and reject a second release of
        a resource that is already back in ``_available``.
        """

        with self._not_empty:
            identity = id(connection)
            if identity not in self._all_ids:
                raise InvalidResourceError("Connection does not belong to this pool")
            if identity not in self._borrowed_ids:
                raise InvalidResourceError("Connection was not borrowed or was already released")
            self._borrowed_ids.remove(identity)
            self._available.append(connection)
            self._not_empty.notify()

    @property
    def size(self) -> int:
        return self._max_size

    @property
    def available(self) -> int:
        with self._lock:
            return len(self._available)

    @property
    def borrowed(self) -> int:
        with self._lock:
            return len(self._borrowed_ids)

File: salesforce/lld/connection-pool/java/src/main/java/com/example/pool/service/ConnectionPool.java
package com.example.pool.service;

import com.example.pool.exception.InvalidResourceException;
import com.example.pool.exception.PoolTimeoutException;
import com.example.pool.factory.ConnectionFactory;
import com.example.pool.factory.InMemoryConnectionFactory;
import com.example.pool.model.Connection;
import com.example.pool.strategy.DefaultValidationStrategy;
import com.example.pool.strategy.ValidationStrategy;

import java.time.Duration;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Set;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Bounded, thread-safe resource pool.
 *
 * <p>The MVP eagerly creates {@code maxSize} fake connections up front, so capacity is fixed and
 * the pool can never over-allocate. Free resources live in a bounded {@link LinkedBlockingQueue};
 * callers of {@link #borrow(Duration)} block on that queue until another thread releases a resource
 * or the timeout expires.
 *
 * <p>Two identity-based sets guard ownership: {@code allConnections} rejects foreign resources,
 * while {@code borrowedConnections} rejects double release. They are protected by one small lock so
 * check-and-mark operations are atomic without blocking the queue itself.
 */
public class ConnectionPool {

    private final int maxSize;
    private final LinkedBlockingQueue<Connection> availableConnections;
    private final Set<Connection> allConnections;
    private final Set<Connection> borrowedConnections;
    private final Object ownershipLock = new Object();
    private final ValidationStrategy validationStrategy;

    public ConnectionPool(int maxSize) {
        this(maxSize, new InMemoryConnectionFactory(), new DefaultValidationStrategy());
    }

    public ConnectionPool(int maxSize,
                          ConnectionFactory factory,
                          ValidationStrategy validationStrategy) {
        if (maxSize <= 0) {
            throw new IllegalArgumentException("maxSize must be positive");
        }
        this.maxSize = maxSize;
        this.availableConnections = new LinkedBlockingQueue<>(maxSize);
        this.allConnections = Collections.newSetFromMap(new IdentityHashMap<>());
        this.borrowedConnections = Collections.newSetFromMap(new IdentityHashMap<>());
        this.validationStrategy = validationStrategy;

        for (int i = 1; i <= maxSize; i++) {
            Connection connection = factory.create("conn-" + i);
            allConnections.add(connection);
            availableConnections.add(connection);
        }
    }

    /**
     * Borrow a resource, waiting up to {@code timeout} for one to become free.
     *
     * <p>{@code poll(timeout)} provides the blocking behavior. Once a connection is removed from the
     * queue, no other caller can receive it. We then mark it as borrowed under {@code ownershipLock}
     * and run the validation strategy before handing it to the client.
     */
    public Connection borrow(Duration timeout) {
        try {
            Connection connection = availableConnections.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
            if (connection == null) {
                throw new PoolTimeoutException("Timed out waiting for a connection");
            }
            synchronized (ownershipLock) {
                borrowedConnections.add(connection);
            }
            if (!validationStrategy.isValid(connection)) {
                release(connection);
                throw new InvalidResourceException("Borrowed connection is not valid: " + connection.getId());
            }
            return connection;
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new PoolTimeoutException("Interrupted while waiting for a connection");
        }
    }

    /**
     * Return a borrowed resource to the pool.
     *
     * <p>The identity checks make ownership explicit: a newly-created/foreign connection is rejected,
     * and a connection already returned once is rejected as a double release.
     */
    public void release(Connection connection) {
        synchronized (ownershipLock) {
            if (!allConnections.contains(connection)) {
                throw new InvalidResourceException("Connection does not belong to this pool");
            }
            if (!borrowedConnections.remove(connection)) {
                throw new InvalidResourceException("Connection was not borrowed or was already released");
            }
        }
        availableConnections.offer(connection);
    }

    /** Configured capacity, not the current number of free resources. */
    public int size() {
        return maxSize;
    }

    /** Current free-resource count (a snapshot under concurrent load). */
    public int available() {
        return availableConnections.size();
    }

    public int borrowed() {
        synchronized (ownershipLock) {
            return borrowedConnections.size();
        }
    }

    public static ConnectionPool withCountingFactory(int maxSize, AtomicInteger createdCount) {
        return new ConnectionPool(maxSize, id -> {
            createdCount.incrementAndGet();
            return new Connection(id);
        }, new DefaultValidationStrategy());
    }
}
